File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int func(vector<int>& weights, int capacity,int days){
        int count=1;
        int sum=0;
        for(int i=0;i<weights.size();i++){
            if(sum+weights[i]>capacity){
                sum=0;
                count++;
            }
            sum+=weights[i];
        }
        return count;
    }
    int shipWithinDays(vector<int>& weights, int days) {
    int low=*max_element(weights.begin(),weights.end());
    int high=accumulate(weights.begin(),weights.end(),0);
    int ans=high;
    while(low<=high){
        int mid=(low+high)/2;
        int count=func(weights,mid,days);
        if(count<=days){
            ans=mid;
            high=mid-1;
        }
        else{
            low=mid+1;
        }
    }
    return ans;
    }
};

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
class Solution:
    def ship_capacity(self,weights,min_weight,days):
        sum=0
        day_count=1
        for i in range(0,len(weights)):
            if(weights[i]+sum<=min_weight):
                sum+=weights[i]
            else:
                sum=weights[i]
                day_count+=1
        return day_count<=days

    def shipWithinDays(self, weights: List[int], days: int) -> int:
        low=max(weights)
        high=sum(weights)
        res=-1
        while(low<=high):
            mid=(low+high)//2
            if(self.ship_capacity(weights,mid,days)):
                res=mid
                high=mid-1
            else:
                low=mid+1
        return res